x/btcstaking/keeper/power_dist_change.go

```diff
@@ -236,6 +237,17 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 		}
 	}
 
+	// set voting power to 0 if the fp does not have timestamped pub rand
+	for _, fp := range newDc.FinalityProviders {
+		// TODO calling HasTimestampedPubRand potentially iterates
+		// all the pub rand committed by the fp, which might slow down
+		// the process, need optimization
+		if fp.TotalVotingPower > 0 &&
+			!k.FinalityKeeper.HasTimestampedPubRand(ctx, fp.BtcPk, height) {
+			fp.TotalVotingPower = 0
+		}
```

Comment:
I just realised that instead of zeroising total voting power of FP to be 0, we would have to remove the FP out from `newDc`, such that `ApplyActiveFinalityProviders` won't count FPs with 0 voting power in the active set when there are very few FPs. This is also done for slashed FPs ([ref](https://github.com/babylonlabs-io/babylon/blob/5a4d7ddae3a93c8e660e8f03fba740538b9e9c6c/x/btcstaking/keeper/power_dist_change.go#L173-L176)). So it seems that https://github.com/babylonlabs-io/babylon/blob/5a4d7ddae3a93c8e660e8f03fba740538b9e9c6c/x/btcstaking/keeper/power_dist_change.go#L173-L176 is the best and the only needed place to filter out FPs without timestamped pub rand. Wdyt? sorry for being slow here

Reply:
Thanks for the reminder! Agree that we should not count fps with 0 voting power in the active set, but I don't think we should remove those who have delegations but don't have timestamped pub rand from the dist cache. If we do so, what if those fps have pub rand suddenly have timestamped pub rand in the next height? Seems we would lose the accumulated delegations to calculate the power in a future height. I think we should fix `GetNumActiveFPs` to return the number of fps that have positive voting power. Did I miss anything?

Reply:
Oh good point. Then indeed the best place to fix this is `ApplyActiveFinalityProviders`. Agree that `GetNumActiveFPs` is probably the place as this function only returns the number of active FPs, although the function name becomes a bit confusing after this change. So how about we 

- refactor `GetNumActiveFPs` as a private keeper function that is only triggered once (It iterates over FPs in DB so is expensive)
- persist the number of active FPs returned by this private keeper function in the `VotingPowerDistCache`
- In `GetActiveFinalityProviderSet` we retrieve `numActiveFPs` from the `dc` object directly, rather than calling the private keeper function

Reply:
I feel that we should not change the state in a getter function like `GetNumActiveFPs`. We should have a separate method `updateNumActiveFps` for that. Please see the update in [8e14b99](https://github.com/babylonlabs-io/babylon/pull/23/commits/8e14b994a747da7ec5892d985b319b26600c7924)

Reply:
I was saying that instead of having this getter function, we can define a new private keeper function (say `calcNumActiveFPs(dc *VotingPowerDistCache)`) that invokes `HasTimestampedPubRand` and determines the number of active FPs, and then use this function to get `num_active_fps` and assign it to `dc`. 

The reasoning was to avoid setting `TotalVotingPower` to zero as we do in L239-247. It's possible that an FP does not have timestamped pub rand at current height but has one in the next height. In this case the `newDc` will give 0 voting power to this FP and the `newDc` is stored in DB. This will be carried to the next height and this FP won't have voting power, which is not expected. It's also a bit hacky as this mixed the concept of voting power and active FP (a FP out of the active FP set can still have positive voting power).

Reply:
Per discussion off-line, we can't set the voting power directly to zero as when the fp has timestamped pub rand in a future height, the voting power can't be set back. We need to figure out another way to do it

Reply:
oh why it can't be set back ? 

I imagine it could be done by iterating over all `BtcDels` and summing once again, though probably the problem would be when to do that. (probably it would require some kind of event for new registred pub rand though which seems like big cahnge)

Reply:
> I imagine it could be done by iterating over all BtcDels and summing once again, though probably the problem would be when to do that.

I think we intentionally avoid doing this due to the overhead

Reply:
I was thinking we could solve the issue by adding a flag `is_timestamped` to the fp dist info. We define that an fp is active at a height as long as it has positive voting power and timestamped pub rand (voting_power > 0 and is_timestamped = true). The voting power in the dist info keeps it is but we don't set the voting power to the store if `is_timestamped=true`. This way, when the fp has timestamped pub rand in a future height, we don't need to re-calculate previous delegations. Please see the change in [b47d837](https://github.com/babylonlabs-io/babylon/pull/23/commits/b47d83735bbc13cbc56ad297a5ee06369b79c21f)